rclcpp/include/rclcpp/node_interfaces/node_interfaces.hpp

```diff
@@ -0,0 +1,127 @@
+// Copyright 2022 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__NODE_INTERFACES__NODE_INTERFACES_HPP_
+#define RCLCPP__NODE_INTERFACES__NODE_INTERFACES_HPP_
+
+#include <memory>
+
+#include "rclcpp/node_interfaces/node_interfaces_helpers.hpp"
+
+namespace rclcpp
+{
+namespace node_interfaces
+{
+
+/// A helper class for aggregating node interfaces
+template<typename ... InterfaceTs>
+class NodeInterfaces
+  : public std::enable_shared_from_this<NodeInterfaces<InterfaceTs...>>, public InterfaceTs ...
+{
+  static_assert(0 != sizeof ...(InterfaceTs), "Template parameters must be populated!");
+
+public:
+  RCLCPP_SMART_PTR_DEFINITIONS(NodeInterfaces)
+
+  /// Create a new node handle with no bound node interfaces.
+  NodeInterfaces()
```

Comment:
the comment seems to refer to the old name (`node handle`). this applies also to other comments below.

moreover, why do we need this constructor?

Reply:
This is necessary for creating an unbound object to then use to aggregate arbitrary sets of node interfaces using the setters!

So, e.g.

Make empty base and clock Interfaces object
Set base to some mock base object
Set clock to some node's clock object

Etc.

Reply:
https://github.com/ros2/rclcpp/compare/1cac3b88b72d6ff02f8aeb71e788d0562a047408..aafce8ca04e6b761f7fcd3d7c77ca08965afea83

Reply:
As I said above, I would prefer no setters and just require people to pass aggregate parts via a constructor.

Reply:
Okay, I'll make this change, the getter question is still open though

Reply:
I have it working locally, though I'd like to highlight that aggregating at construction means we're going to see some really janky looking formatting (as opposed to using setters).

```
    auto base_interfaces = std::make_shared<rclcpp::node_interfaces::NodeInterfaces<
          rclcpp::node_interfaces::Base,
          rclcpp::node_interfaces::Clock,
          rclcpp::node_interfaces::Graph
        >>(
      node->get_node_base_interface(),
      node->get_node_clock_interface(),
      node->get_node_graph_interface()
        );
```

Reply:
I don't think making this class mutable just for style is the right approach, and you can make the formatting better, e.g.:

```c++
  using MyBaseNodeInterfaces = rclcpp::node_interfaces::NodeInterfaces<
    rclcpp::node_interfaces::Base,
    rclcpp::node_interfaces::Clock,
    rclcpp::node_interfaces::Graph
  >;
  auto base_interfaces = std::make_shared<MyBaseNodeInterfaces>(
    node->get_node_base_interface(),
    node->get_node_clock_interface(),
    node->get_node_graph_interface());
```

I ran that through `ament_uncrustify`.

Reply:
If you still want to have a iterative approach to creating it, then I would say adding a "builder" pattern would be better than making it mutable, something like:

```c++
  auto base_interfaces_builder  = rclcpp::node_interfaces::NodeInterfacesBuilder<
    rclcpp::node_interfaces::Base,
    rclcpp::node_interfaces::Clock,
    rclcpp::node_interfaces::Graph
  >();
  auto base_interfaces = base_interfaces_builder
    .node_base(node->get_node_base_interface())
    .node_clock(node->get_node_clock_interface())
    .node_graph(node->get_node_graph_interface())
    .build();
```

Or some version of that. Though I think for the added amount of code to test and maintain in rclcpp it isn't worth it. However, I think it is preferable to making the class mutable, because if it is mutable it makes everything using it have to be more considerate, since it could change at any point, making thread-safe code more difficult. However, if you know it's immutable then you can just access it without much worry.

Also, if there were use cases for mutating it after creation then making it mutable would make more sense I think, but as far as I can imagine you'd only really use this when constructing it, which is exactly what the builder pattern is used for.

Reply:
Fair point! The next iteration is ready!